### ISSUE TITLE
How can I load MBTiles(with vectors) data into TileDB? 

### ISSUE THREAD

**Issue by OrlovDiga** [TileDB-Inc/TileDB-Geospatial#15]

**Comment by normanb** [comment]
@OrlovDiga Raster MBTiles can be loaded as can MBTiles with points.

Raster MBTiles can be converted using GDAL (or Rasterio) to TileDB arrays.

There is a GDAL OGR driver in development for MBTiles with points but for now you would have to use our [TileDB-Py](https://github.com/TileDB-Inc/TileDB-Py) library directly. 

Do you have a dataset that you can share and we can create an example.

**Comment by OrlovDiga** [comment]
@normanb thank you very much for your answer. Yes, you can take the sample data (MBTiles for vectors) from this [resource](https://openmaptiles.com/) ([Example small data](https://openmaptiles.com/downloads/russia/chelyabinsk/))

**Comment by normanb** [comment]
@OrlovDiga I hit a paywall with that link, do you have another example?

**Comment by OrlovDiga** [comment]
Hello @normanb, I found another example - there is a link in the readme [link](https://github.com/klokantech/vector-tiles-sample)

**Comment by normanb** [comment]
@OrlovDiga I explored this by running `https://github.com/mapbox/mbutil` to dump the mbtile to disk in pbf format e.g. `mb-util countries.mbtiles --image_format=pbf tiles`, it is possible to read the mbtile directly however this just made it easier to understand. 

The mbtile file structure is a collection of mapbox vector tiles (MVT) in protobuf (as vector) format laid out by zoom level. This would correspond to a TileDB Group. Within each group we could have an array (either sparse or dense) with points that contain the MVT `MOVETO`, `LINETO`, `CLOSEPATH` commands.

What is your use case for doing this? I think it is doable, I would like to better understand how you are going to use this.

**Comment by OrlovDiga** [comment]
@normanb Thank you very much for your answer.
I am writing a geo server using TileDB as geodata storage (raster and vector). my custom geo server will serve geodata to the frontend using Mapbox.

Do you have information on how best to do it? For example, what data format(raster) to use between frontend and backend? Where can I see examples similar to my assignment? Which libraries will help me?

Best regards, OrlovDiga

**Comment by OrlovDiga** [comment]
And yes, with which method at  [TileDB-Py](https://github.com/TileDB-Inc/TileDB-Py) can I convert MbTile to TileDB storage?
In the [docs](https://tiledb-inc-tiledb-py.readthedocs-hosted.com/en/stable/python-api.html), I only found for converting from CSV, NumPy, pandas.

**Comment by normanb** [comment]
@OrlovDiga Thanks, I understand a little better now. It is possible to use TileDB as geodata storage for raster data and then use MapServer, GeoServer or Mapnik to render the data, this is enabled through the GDAL TileDB driver.

For point data and these rendering tools you will need an OGR driver and if you want to do that as a proof of concept with your assignment then you can write a python OGR driver using https://gdal.org/development/rfc/rfc76_ogrpythondrivers.html and the TileDB-Py binding. Feel free to ask questions here and we will help where we can.

**Comment by OrlovDiga** [comment]
@normanb I would like to inform you that at the moment the GeoServer does not support the required TileDB format from GDAL. [link to GeoServer doc](https://docs.geoserver.org/stable/en/user/data/raster/gdal.html)

P.S. Or am I missing something or not understanding? :)

**Comment by normanb** [comment]
@OrlovDiga My understanding is Geoserver will read TileDB arrays if the TIleDB driver is enabled in GDAL. When I Iast tested this it worked. GDAL can be installed from Conda or from source. Let us know if you have any issues with the build and installation of the GDAL TileDB driver.

**Comment by OrlovDiga** [comment]
> @OrlovDiga Thanks, I understand a little better now. It is possible to use TileDB as geodata storage for raster data and then use MapServer, GeoServer or Mapnik to render the data, this is enabled through the GDAL TileDB driver.
> 
> For point data and these rendering tools you will need an OGR driver and if you want to do that as a proof of concept with your assignment then you can write a python OGR driver using https://gdal.org/development/rfc/rfc76_ogrpythondrivers.html and the TileDB-Py binding. Feel free to ask questions here and we will help where we can.

@normanb  Ok, thank you, I'll try to write a driver a little later. What exactly can help me in converting with TileDB-Py from MBTile to TileDB storage?

**Comment by normanb** [comment]
@OrlovDiga For vectors you will need to read the individual mapbox vector tiles that are contained in the MBTile package.

I dumped the mbtile you linked to a tree of protobufs (PBF) using the Python `mb-util` package e.g. `mb-util countries.mbtiles --image_format=pbf tiles` 

From there you will want to use `TileDB-Py` to create a TileDB group and an array for each zoom level with an attribute that will contain the `MOVETO`, `LINETO` instructions so that as you parse each of the input files you write the drawing instructions.  Once you have the vectors in the TileDB array you will then need to write a OGR driver for integration with GeoServer. 

MBTiles with rasters works for translating to TileDB arrays right now with GDAL, but we don't support MBTiles with points in any of our TileDB integrations though it is possible with the method I outlined above.
